django/db/models/query.py

```diff
@@ -2448,14 +2482,22 @@ def prefetch_one_level(instances, prefetcher, lookup, level):
     # The 'values to be matched' must be hashable as they will be used
     # in a dictionary.
 
+    # If the lookup is a Generic Prefetch class,
+    # it should send the querysets argument to get_prefetch_queryset
+    prefetch_queryset_kwargs = (
+        {"querysets": lookup.get_current_querysets(level)}
+        if isinstance(lookup, GenericPrefetch)
+        else {"queryset": lookup.get_current_queryset(level)}
+    )
+
     (
         rel_qs,
         rel_obj_attr,
         instance_attr,
         single,
         cache_name,
         is_descriptor,
-    ) = prefetcher.get_prefetch_queryset(instances, lookup.get_current_queryset(level))
+    ) = prefetcher.get_prefetch_queryset(instances, **prefetch_queryset_kwargs)
```

Comment:
This seems like a code smell to me; the core should not know about the existence of `GenericPrefetch`.

It's not a big deal if the kwarg and methods are named `get_current_queryset` and `queryset` IMO.

Reply:
@charettes  Having `queryset` instead of `querysets` is semantically wrong. Good approach might be to have two generic functions `get_current_querysets` and `get_current_queryset` in the `Prefetch` class to populate the two arguments `queryset` and `querysets`. Function `get_current_querysets` in the `Prefetch` class can return `None`

This way core need not know about about the lookup and it can use the functions `get_current_querysets` and `get_current_queryset` to set kwargs `querysets` and `queryset` respectively. What do think?

Reply:
I get what you mean.

Another approach would be to deprecate `Prefetch.get_current_queryset` entirely and only implement `get_current_querysets` while keeping `get_current_queryset` as a shim to raise a deprecation warning.

The same could then be done for `get_prefetch_queryset` which would be deprecated in favor of `get_prefetch_querysets`?

Reply:
Yeah, deprecating and replacing with `get_prefetch_querysets` would give a lot more options to extend current lookups.

I will change the PR so that the `get_prefetch_queryset` is deprecated and make it backward compatible.